src/lib.rs

```diff
@@ -0,0 +1,163 @@
+use std::ffi::{c_void, CStr, CString};
+use std::io::{BufRead, Error, Read};
+use std::os::raw::{c_char, c_uchar, c_uint};
+use std::ptr::{null, null_mut};
+
+use crate::go::slice::{GoSlice, ToGoSlice};
+use crate::go::string::{GoString, ToGoString};
+use crate::go::guess::{GoGuess, Guess};
+
+mod go;
+
+extern "C" {
+    fn GetLanguage(filename: GoString, content: GoSlice) -> GoString;
+    fn GetMimeType(path: GoString, language: GoString) -> GoString;
+    fn GetLanguages(filename: GoString, content: GoSlice, result: &mut GoSlice);
+    fn GetLanguageExtensions(language: GoString, result: &mut GoSlice);
+
+    fn GetLanguageByContent(filename: GoString, content: GoSlice) -> GoGuess;
+    fn GetLanguageByExtension(filename: GoString) -> GoGuess;
+    fn GetLanguageByFilename(filename: GoString) -> GoGuess;
+    fn GetLanguageByModeline(content: GoSlice) -> GoGuess;
+    fn GetLanguageByShebang(content: GoSlice) -> GoGuess;
+    fn GetLanguageByVimModeline(content: GoSlice) -> GoGuess;
+    fn GetLanguageByEmacsModeline(content: GoSlice) -> GoGuess;
+
+    fn IsBinary(data: GoSlice) -> c_uchar;
+    fn IsConfiguration(path: GoString) -> c_uchar;
+    fn IsDocumentation(path: GoString) -> c_uchar;
+    fn IsDotFile(path: GoString) -> c_uchar;
+    fn IsImage(path: GoString) -> c_uchar;
+    fn IsVendor(path: GoString) -> c_uchar;
+    fn IsGenerated(path: GoString, content: GoSlice) -> c_uchar;
+    fn GetColor(language: GoString) -> GoString;
+
+    // Additional internal functions possible to define
+    // fn GetLanguagesByContent(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+    // fn GetLanguagesByEmacsModeline(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+    // fn GetLanguagesByExtension(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+    // fn GetLanguagesByFilename(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+    // fn GetLanguagesByModeline(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+    // fn GetLanguagesByShebang(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+    // fn GetLanguagesByVimModeline(filename: GoString, content: GoSlice, candidates: GoSlice, result: &mut GoSlice);
+}
+
+pub fn get_languages(filename: &str, content: &str) -> Vec<String> {
+    let c_filename = CString::new(filename).expect("Can't construct string");
```

Comment:
Why it this handled differently than in `get_language_extensions` and others?

Reply:
Thanks, nice catch. Looks like there should be unwrap, because user would have nothing to do if there is some kind of error with building C strings from rust's ones.

Reply:
Hmm, after some playing with the bindings, I think that's more appropriate way to handle this is to pass the error down to handle by user. Will fix this and all the functions to return `Result` instead of just String.